pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -0,0 +1,100 @@
+package virtualmachine
+
+import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubevirtapis "kubevirt.io/client-go/api/v1alpha3"
+
+	cdictrl "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
+)
+
+type VMController struct {
+	// dv ctrl suite
+	dataVolumeClient cdictrl.DataVolumeClient
+	dataVolumeCache  cdictrl.DataVolumeCache
+}
+
+// RefOutTreeDataVolumes records the target VirtualMachine as the owner of the out-tree DataVolumes.
+func (h *VMController) RefOutTreeDataVolumes(_ string, vm *kubevirtapis.VirtualMachine) (*kubevirtapis.VirtualMachine, error) {
+	if vm == nil || vm.DeletionTimestamp != nil || vm.Spec.Template == nil {
+		return vm, nil
+	}
+
+	var outTreeDataVolumeNames = getOutTreeDataVolumeNames(&vm.Spec)
+
+	// adopts orphan out-tree DataVolumes
+	var dataVolumeNamespace = vm.Namespace
+	for _, dataVolumeName := range outTreeDataVolumeNames {
+		var dataVolume, err = h.dataVolumeCache.Get(dataVolumeNamespace, dataVolumeName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// NB(thxCode): ignores error, since this can't be fixed by an immediate requeue,
```

Comment:
requeue -> `re-sync/re-queue`

Reply:
I think `requeue` is enough: https://github.com/kubernetes-sigs/controller-runtime/search?q=requeue&unscoped_q=requeue.